esp/services/esdl_svc_engine/esdl_svc_custom.cpp

```diff
@@ -0,0 +1,331 @@
+            continue;
+        }
+
+        const char * pathToValue = cur.queryProp("hpcc:Value/@path");
+        if (!pathToValue || !*pathToValue)
+        {
+            //if(required)
+               //throw??
+            //else
+            ERRLOG("Encountered SetValue without Value path");
+            continue;
+        }
+
+        addTransformRule(new CEsdlCustomTransformRule(ruleName, targetFieldPath, pathToValue, required, typeIsSet ), otherwise);
+    }
+}
+
+void CEsdlCustomTransformClause::compileRules(IPropertyTree * rules, bool otherwise)
+{
+    compileRules(rules->getElements("hpcc:SetValue"), otherwise, true);
+    compileRules(rules->getElements("hpcc:AppendValue"), otherwise, false);
+}
+
+void CEsdlCustomTransformClause::addTransformRule(CEsdlCustomTransformRule * fieldmapping, bool otherwise)
+{
+    if (!otherwise)
+        m_transformRules.append(*LINK(fieldmapping));
+    else
+        m_otherwiseTransformRules.append(*LINK(fieldmapping));
+}
+
+void CEsdlCustomTransformClause::addNestedTransformClause(CEsdlCustomTransformClause * nestedConditional, bool otherwise)
+{
+    if(!otherwise)
+        m_nestedTransformClauses.append(*LINK(nestedConditional));
+    else
+        m_otherwiseNestedTransformClauses.append(*LINK(nestedConditional));
+}
+
+bool CEsdlCustomTransformClause::evaluate(IEspContext * context, IPropertyTree *request, IPropertyTree * bindingCfg)
+{
+    StringBuffer xmlreq;
+    toXML(request, xmlreq);
+
+    m_xpathContext->setXmlDoc(xmlreq.str());
+
+    VStringBuffer ver("%g", context->getClientVersion());
+    if(!m_xpathContext->addParameter("clientversion", ver.str()))
+        ERRLOG("Could not set custom transform variable: clientversion:'%s'", ver.str());
+
+    auto user = context->queryUser();
+    if (user)
+    {
+        auto userPropIt = user->getPropertyIterator();
+
+        ForEach(*userPropIt)
+        {
+            const char *name = userPropIt->getPropKey();
+            if (name && *name)
+                m_xpathContext->addParameter(name, user->getProperty(name));
+        }
+    }
+
+    Owned<IPropertyTreeIterator> configParams = bindingCfg->getElements("Transform/Param");
+    {
+        ForEach(*configParams)
+        {
+            IPropertyTree & currentParam = configParams->query();
+            m_xpathContext->addParameter(currentParam.queryProp("@name"), currentParam.queryProp("@value"));
+        }
+    }
+
+    bool evalresp = false;
+    try
+    {
+        evalresp = m_xpathContext->evaluateAsBoolean(m_compiledConditionalXpath.getLink());
+    }
+    catch (...)
+    {
+        DBGLOG("CEsdlTransformConditional:evaluate: Could not evaluate xpath '%s'", m_xpathContext->getXpath());
+    }
+    return evalresp;
+}
+
+void CEsdlCustomTransformClause::process(IEspContext * context, IPropertyTree *request, IPropertyTree * bindingCfg)
+{
+    bool result = false;
+    try
+    {
+        bool result = evaluate(context, request,bindingCfg);
```

Comment:
I could use a little explanation of this arrangement where you evaluate and then pass the result to various types of children.  Might be ok, but a little confusing to me.  Let's discuss offline, then I'll comment more here.

Reply:
@rpastrana